src/rules/no-useless-path-segments.js

```diff
@@ -45,65 +62,99 @@ module.exports = {
         type: 'object',
         properties: {
           commonjs: { type: 'boolean' },
+          detectUselessIndex: { type: 'boolean' },
         },
         additionalProperties: false,
       },
     ],
 
     fixable: 'code',
+    messages: {
+      uselessPath: 'Useless path segments for "{{ path }}", should be "{{ proposedPath }}"',
+    },
   },
 
-  create: function (context) {
+  create(context) {
     const currentDir = path.dirname(context.getFilename())
+    const options = context.options[0]
 
     function checkSourceValue(source) {
-      const { value } = source
+      const { value: importPath } = source
 
-      function report(proposed) {
+      function report(proposedPath) {
         context.report({
           node: source,
-          message: `Useless path segments for "${value}", should be "${proposed}"`,
-          fix: fixer => fixer.replaceText(source, JSON.stringify(proposed)),
+          messageId: 'uselessPath',
+          data: {
+            path: importPath,
+            proposedPath,
+          },
+          fix: fixer => proposedPath && fixer.replaceText(source, JSON.stringify(proposedPath)),
         })
       }
 
-      if (!value.startsWith('.')) {
+      // Only relative imports are relevant for this rule --> Skip checking
+      if (!importPath.startsWith('.')) {
         return
       }
 
-      const resolvedPath = resolve(value, context)
-      const normed = normalize(value)
-      if (normed !== value && resolvedPath === resolve(normed, context)) {
-        return report(normed)
+      // Report rule violation if path is not the shortest possible
+      const resolvedPath = resolve(importPath, context)
+      const normedPath = normalize(importPath)
+      const resolvedNormedPath = resolve(normedPath, context)
+      if (normedPath !== importPath && resolvedPath === resolvedNormedPath) {
+        return report(normedPath)
+      }
+
+      // Path contains unnecessary /index
+      if (options && options.detectUselessIndex && importPath.endsWith('index')) {
```

Comment:
should this be checking importPath, or something else? what happens if they’ve specified `index.js`?

Reply:
What do you mean by "should this be checking importPath, or something else?"?

I've added a test for `../index.js`. Importing something like `../index.js` doesn't cause any warnings. The [extensions](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/extensions.md) rule can find those issues. Apparently, the `extension` rule is not able to auto-fix violations. Fixing `../index.js` to be `../` would require 2 steps if both rules are enabled.

Reply:
That means that if a user has `.js` required, they can't get `index.js` auto-removed?

Reply:
Yes, exactly. They would get a warning but the `extension` rule currently is not able to auto-fix those things.

Reply:
It seems like the extensions rule should be unrelated here; any path that could be shortened should be caught by this rule.

Reply:
Ok, I'll add this feature then. Would you consider it part of the `detectUselessIndex` option or add a new option for this?

Reply:
I've updated the PR to find and fix `index.js` as well (depending on the configured file extensions, other file extensions would be found, too). I haven't added a new option for this. Let me know if you prefer an additional option.

Reply:
I think it should be part of `detectUselessIndex`, yes.